### PR TITLE
[UXIT-1507] Add ecosystem project submit form

### PR DESCRIPTION
**Pull request by CharlyMartin**
## 📝 Description

This PR introduces a form for ecosystem project owners to submit their information for inclusion on our website. The form is available at `/ecosystem-explorer/project-form`

**New feature**

## 🛠️ Key Changes

- Adds a new page at `/ecosystem-explorer/project-form` using the generate page script.
- Creates a folder for ecosystem page content at `src/content/pages/ecosystem-explorer/*md`.
- Adds all form-related code (components, utils, constants, hooks, and schema) in `app/ecosystem-explorer/project-form`.
- Updates `ControlledFormFileInput.tsx` to accept and return `Array<File>` instead of `FileList`. This simplifies form initialization by allowing an empty array instead of using `new FileList`, which is not available server-side.
- Updates `ControlledFormFileInput.tsx` to accept `undefined` for optional values.
- Updates `FormLabel.tsx` to accept an `optional` prop, to display _(optional)_ next to the form label
- Removes the Airtable link from the `/ecosystem-explorer` page header and links it to the new form.

## 🧪 How to Test

1. Go to /ecosystem-explorer/project-form.
2. Fill out and submit the form.
3. Verify that the created PR contains all the submitted information in the correct format.
4. Open the preview link and confirm that the ecosystem project is displayed with the correct format.

## 📸 Screenshots

CleanShot 2024-09-12 at 11 48 11@2x

### CODE REVIEW

**Comment by barbaraperic** [comment]
Text overflows on `md` screen size. [According to Figma, it should be aligned with the form](https://www.figma.com/design/0pBfQs5tKI6bx5ypPIfVRA/FIL.org-V4-Prototypes?node-id=1231-24946&t=pOCk7P5UZf5mhww8-4)?

Screenshot 2024-09-12 at 12 42 17

**Comment by barbaraperic** [comment]
This is not a part of this PR, but we should probably open a ticket to take a look at the `PageHeader` component - gap should be 64px but the margin adds up? 

Screenshot 2024-09-12 at 12 42 55

File: src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+              background, at least 1000px by 1000px, and under 100KB. You can
+              use tools like{' '}
+              <a
+                href="https://squoosh.app/"
+                className="text-brand-300 underline hover:text-brand-400"
+                rel="noopener noreferrer"
+                target="_blank"
+                aria-label='Opens "Sqoosh" in a new tab'
+              >
+                Sqoosh
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx:174]:
```suggestion
                Squoosh
```

File: src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            <>
+              For best quality, please submit a white logo with a transparent
+              background, at least 1000px by 1000px, and under 100KB. You can
+              use tools like{' '}
+              <a
+                href="https://squoosh.app/"
+                className="text-brand-300 underline hover:text-brand-400"
+                rel="noopener noreferrer"
+                target="_blank"
+                aria-label='Opens "Sqoosh" in a new tab'
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx:172]:
```suggestion
                aria-label='Opens "Squoosh" in a new tab'
```

**Comment by barbaraperic** [comment]
Twitter Handle error message

Screenshot 2024-09-12 at 13 02 41

**Comment by barbaraperic** [comment]
Should Submit project button be disabled if the form is empty?

https://github.com/user-attachments/assets/22fb86bd-cc8c-4ab1-92cc-b7b38c6ebe94

**Comment by barbaraperic** [comment]
Both can be selected? Or is it only one technology?

Screenshot 2024-09-12 at 13 42 39

File: src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+          name="briefSummary"
+          label="Brief Summary"
+          placeholder="Write a brief summary of your project..."
+          characterLimit={BRIEF_CHARACTER_LIMIT}
+          disabled={isSubmitting}
+        />
+
+        <ControlledFormTextarea<EcosystemProjectFormData>
+          name="networkUseCase"
+          label="Network Use Case"
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx:129]: "How do you use the Filecoin Network?"? [Figma](https://www.figma.com/design/0pBfQs5tKI6bx5ypPIfVRA/FIL.org-V4-Prototypes?node-id=1231-24804&t=CKzEG4OJYmUxQYtb-4)

Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/ProjectForm.tsx:129]:
```suggestion
          label="How do you use the Filecoin Network?"
```

 [Figma](https://www.figma.com/design/0pBfQs5tKI6bx5ypPIfVRA/FIL.org-V4-Prototypes?node-id=1231-24804&t=CKzEG4OJYmUxQYtb-4)

**Comment by barbaraperic** [comment]
For the dropdown select - should we enable scrolling? Also if you try to use the arrow keys, it does weird hops between categories:


https://github.com/user-attachments/assets/1ede7a2f-d710-451b-88de-850dc240b23f

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+              label="Filecoin"
+              disabled={isSubmitting}
+            />
+            <ControlledFormCheckbox<EcosystemProjectFormData>
+              name="tech.ipfs"
+              label="IPFS"
+              disabled={isSubmitting}
+            />
+          </div>
+        </FormField>
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx:108]: We get this error in the browser - I manage to isolate it to this part, tho I don't know what is causing it.

Screenshot 2024-09-12 at 14 39 44

File: src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          format: getFileFormat(logo.name),
+        },
+        category: data.category.id,
+        subcategories: [data.topic.id],
+        tech: keepTruthyKeysInArray(data.tech),
+        shortDescription: data.briefSummary,
+        longDescription: data.networkUseCase,
+        yearJoined: data.yearJoined.name,
+        websiteUrl: data.websiteUrl,
+        youtubeUrl: formatYoutubeEmbedUrl(data.youtubeUrl),
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts:33]: Super cool that you included embedUrl ✨

File: src/app/ecosystem-explorer/project-form/schema/form.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  tech: z
+    .object({
+      filecoin: z.boolean(),
+      ipfs: z.boolean(),
+    })
+    .refine(validateOneCheckboxSelected, {
+      message: 'Select at least one technology',
+      path: ['root'],
+    }),
+  yearJoined: ObtionSchema,
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/schema/form.ts:41]:
```suggestion
  yearJoined: OptionSchema,
```

File: src/app/ecosystem-explorer/project-form/schema/form.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  TWITTER_BASE_URL,
+  X_BASE_URL,
+  YOUTUBE_BASE_URL,
+} from '../constants'
+
+const invalidUrlMessage = { message: 'Invalid URL format' } as const
+
+const EmptyStringSchema = z.literal('')
+
+const ObtionSchema = z
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/schema/form.ts:19]:
```suggestion
const OptionSchema = z
```

File: src/app/ecosystem-explorer/project-form/schema/form.ts [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    name: z.string(),
+  })
+  .refine(validateListboxSelected, {
+    message: 'Please select an option',
+  })
+
+export const EcosystemProjectFormSchema = z.object({
+  name: z.string().min(1, { message: 'Your name is required' }),
+  email: z.string().email({ message: 'The email format is invalid' }),
+  projectName: z.string().min(1, { message: 'Your name is required' }),
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/schema/form.ts:31]:
```suggestion
  projectName: z.string().min(1, { message: 'Your project name is required' }),
```

File: src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        githubUrl: data.githubUrl,
+        xHandle: data.xHandle,
+        timestampISO: new Date().toISOString(),
+      })
+      updateSearchParams({ status: 'success', prNumber: pullRequest.number })
+    } catch (error) {
+      updateSearchParams({ status: 'error', message: String(error) })
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts:40]: Maybe we can use custom errors, something like "Something went wrong, try again or email ..."

Comment (barbaraperic) [reply src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts:40]: + Sentry catch error

File: src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    }
+  }
+}
+
+function keepTruthyKeysInArray(object: Record<string, boolean>) {
+  return Object.entries(object)
+    .filter(([, value]) => value)
+    .map(([key]) => key)
+}
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts:49]: Took me a minute to figure out what it does, but very cool 👌 Now when I read it again, makes sense

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        <ControlledFormInput<EcosystemProjectFormData>
+          name="websiteUrl"
+          label="Website URL"
+          placeholder="Website URL"
+          type="url"
+          disabled={isSubmitting}
+        />
+
+        <ControlledFormInput<EcosystemProjectFormData, StringOrUndefined>
+          optionalLabel
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx:190]: First time I read it it sounded like the label is optional? But actually the field is optional? Wondering if should rename the prop to optional? What do you think?

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+          disabled={isSubmitting}
+        />
+      </FormSection>
+      <Button
+        variant="primary"
+        disabled={isSubmitting}
+        className="w-full sm:w-auto"
+      >
+        {isSubmitting ? 'Submitting...' : 'Submit Project'}
+      </Button>
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx:222]: Think I included it in one of comments when I was testing the form - if we could disable the button when the form is empty?

File: src/app/ecosystem-explorer/project-form/components/SuccessMessage.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+        <Heading tag="h1" variant="4xl">
+          Thank you!
+        </Heading>
+
+        <p>
+          We’ve received your submission and it will now be reviewed by our
+          team. You can follow the progress on GitHub{' '}
+          <a
+            href={`${FILECOIN_FOUNDATION_GITHUB_PR_URL}/${prNumber}`}
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/SuccessMessage.tsx:39]: Maybe we can extract this in a variable and put below line 16? So we have urls grouped together?

File: src/app/ecosystem-explorer/project-form/constants/index.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+export const MAX_FILE_SIZE_IN_BYTES = 100_000
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/constants/index.ts:1]: the typescript number format is 👌 didn't know about it

File: src/app/ecosystem-explorer/project-form/services/github/index.ts [hunk 15]
```diff
@@ -3,125 +3,122 @@
...
+    newTrees: [
+      {
+        path: `${mediaFolder}/${slug}.${data.logo.format}`,
+        mode: '100644',
+        type: 'blob',
+        sha: imageBlob.sha,
+      },
+      {
+        path: `${ecosystemFolder}/${slug}.md`,
+        mode: '100644',
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/services/github/index.ts:87]: save in a constant?

**Review by barbaraperic — COMMENTED** [review]
@CharlyMartin done 😮‍💨 well done, this was a lot of work 👏 I left a few comments, let me know if I need to clarify

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+        <ControlledFormInput<EcosystemProjectFormData>
+          name="name"
+          label="Name"
+          placeholder="Your Name"
+          disabled={isSubmitting}
+        />
+        <ControlledFormInput<EcosystemProjectFormData>
+          name="email"
+          label="Email"
+          placeholder="Your Email"
```
Comment (barbaraperic) [inline src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx:78]: This just says "Email" [in Figma](https://www.figma.com/design/0pBfQs5tKI6bx5ypPIfVRA/FIL.org-V4-Prototypes?node-id=5196-68950&t=bTzI5ZqPyweteofc-4), as a placeholder.

File: src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts (hunk 9 above)
Comment (barbaraperic) [reply src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts:40]: also for debugging purposes? `console.error('Error in submitProjectToGithub:', error)`

**Comment by barbaraperic** [comment]
Mobile view seems to have a lot of space? Also, probably not a part of this PR, but we should fix the breadcrumbs on mobile view...

Screenshot 2024-09-13 at 09 45 53

File: src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts (hunk 10 above)
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/project-form/hooks/useSubmitEcosystemProjectForm.ts:49]: Rewrote it :)

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx (hunk 11 above)
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx]: I agree, renamed it to `addOptionalToLabel`

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx (hunk 12 above)
Comment (barbaraperic) [reply src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx:222]: Ah yes, you're right. I had an impression that it reloaded the page, but I can see now the errors 👍

**Comment by barbaraperic** [comment]
Not sure if this is a big UX issue - or an easy fix - but adding space shows an error. Use case what I can think of is, if the user is copy pasting the url and it has a space in front, they will see an error, tho the url is correct. Also something we can pick up later I guess.

https://github.com/user-attachments/assets/a084fa9a-bc42-4083-a4b1-b37c644c0495

**Comment by barbaraperic** [comment]
Should we limit the `Project Name` character size?

Screenshot 2024-09-13 at 12 59 42

File: src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx (hunk 4 above)
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/project-form/components/EcosystemProjectForm.tsx]: Should be fixed.

**Pull request merged**